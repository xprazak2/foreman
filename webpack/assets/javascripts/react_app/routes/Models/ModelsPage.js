import React from 'react';
import { Button } from 'patternfly-react';
import PageLayout from '../common/PageLayout/PageLayout';
import ModelsTable from '../../components/ModelsTable';

const ModelsPage = props => {
  const { searchProps, pagination, searchable } = props;

  return (
    <PageLayout
      header={__('Hardware Models')}
      searchable={searchable}
      searchProps={searchProps}
      toolbarButtons={
        <Button>Button!</Button>
      }
    >
      <ModelsTable />
    </PageLayout>
  );
}

export default ModelsPage;
