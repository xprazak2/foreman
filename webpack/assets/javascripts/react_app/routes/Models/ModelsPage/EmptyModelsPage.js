import React from 'react';

import DefaultEmptyState from '../../../components/common/EmptyState';
import { foremanUrl } from '../../../../foreman_tools';

const EmptyModelsPage = props => {
  return (
    <DefaultEmptyState
      icon='server'
      iconType='fa'
      header={__('Hardware Models')}
      description={__('Hardware models describe the hardware types of your hosts, including CPU class, vendor class and other notes.')}
      action={{ onClick: () => window.location.href = foremanUrl('/models/new'), title: __('Create Hardware Model') }}
    />
  );
}

export default EmptyModelsPage;
