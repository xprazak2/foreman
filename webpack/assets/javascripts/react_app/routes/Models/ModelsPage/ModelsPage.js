import React from 'react';

import PageLayout from '../../common/PageLayout/PageLayout';
import ModelsPageContent from './components/ModelsPageContent';
import { MODELS_SEARCH_PROPS } from '../constants';

const ModelsPage = props => {
  console.log(props);
  const handleSearch = search => props.fetchAndPush({ searchQuery: search, page: 1 })

  return (
    <PageLayout
      header={__('Hardware Models')}
      searchable
      searchProps={MODELS_SEARCH_PROPS}
      searchQuery={props.search}
      isLoading={props.isLoading && props.hasData}
      onSearch={handleSearch}
      onBookmarkClick={handleSearch}
    >
      <ModelsPageContent
        models={props.models}
        page={props.page}
        perPage={props.perPage}
        search={props.search}
        sort={props.sort}
        hasData={props.hasData}
        hasError={props.hasError}
        isLoading={props.isLoading}
        itemCount={props.itemCount}
        fetchAndPush={props.fetchAndPush}
        message={props.message}
      />
    </PageLayout>
  );
}

export default ModelsPage;
