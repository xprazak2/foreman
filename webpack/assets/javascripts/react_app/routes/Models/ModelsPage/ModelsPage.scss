table.table.table-striped {
  margin-bottom: -1px;
}
