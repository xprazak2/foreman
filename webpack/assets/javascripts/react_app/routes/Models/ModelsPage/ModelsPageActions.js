import history from '../../../history';
import API from '../../../API';

import { deepPropsToCamelCase } from '../../../common/helpers';

import {
  MODELS_PAGE_DATA_RESOLVED,
  MODELS_PAGE_DATA_FAILED,
  MODELS_PAGE_HIDE_LOADING,
  MODELS_PAGE_UPDATE_QUERY,
  MODELS_PAGE_CLEAR_ERROR,
  MODELS_PAGE_SHOW_LOADING,
} from '../constants';

import {
  selectIsLoading,
  selectHasError,
  selectSearch,
  selectPage,
  selectPerPage
} from './ModelsPageSelectors';

const pathname = '/hw_models';

import { stringifyParams, getParams } from '../../../common/urlHelpers';

export const initializeModels = () => dispatch => {
  const params = getParams();
  dispatch(fetchModels(params));
  if (!history.action === 'POP') {
    history.replace({
      pathname: pathname,
      search: stringifyParams(params),
    });
  }
}

export const fetchModels = (
  { page, perPage, searchQuery },
  url = '/api/models?include_permissions=true'
) => async (dispatch, getState) => {
  dispatch({ type: MODELS_PAGE_SHOW_LOADING });
  if (selectHasError(getState())) {
    dispatch({ type: MODELS_PAGE_CLEAR_ERROR });
  }

  const onSuccess = ({ data }) => {
     if (selectIsLoading(getState())) {
      dispatch({ type: MODELS_PAGE_HIDE_LOADING });
    }

    const transformedResponse = deepPropsToCamelCase(data)
    const itemCount = transformedResponse.subtotal

    dispatch({
      type: MODELS_PAGE_DATA_RESOLVED,
      payload: { ...transformedResponse, hasData: itemCount > 0 }
    });
  }

  const onError = error => {
    if (selectIsLoading(getState())) {
      dispatch({ type: MODELS_PAGE_HIDE_LOADING });
    }
    console.log(error);

    dispatch({
      type: MODELS_PAGE_DATA_FAILED,
      payload: {
        message: {
          type: 'error',
           text: `${error.response.status} ${__(error.response.statusText)}`,
        }
      }
    })
  }

  try {
    const response = await API.get(
      url,
      {},
      {
        page,
        per_page: perPage,
        search: searchQuery,
      }
    );
    return onSuccess(response);
  } catch(error) {
    return onError(error);
  }
}

export const fetchAndPush = params => (dispatch, getState) => {
  console.log(params)
  const query = buildQuery(params, getState());
  dispatch(fetchModels(query));
  history.push({
    pathname: pathname,
    search: stringifyParams(query)
  });
}

const buildQuery = (query, state) => ({
  page: query.page || selectPage(state),
  perPage: query.perPage || selectPerPage(state),
  searchQuery: query.searchQuery === undefined
    ? selectSearch(state)
    : query.searchQuery
});
