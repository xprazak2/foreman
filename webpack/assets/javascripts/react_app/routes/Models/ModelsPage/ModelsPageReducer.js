import Immutable from 'seamless-immutable';

import {
  MODELS_PAGE_LOAD_REQUEST,
  MODELS_PAGE_LOAD_ERROR,
  MODELS_PAGE_LOAD_SUCCESS,
} from '../consts';

const initialState = Immutable({
  loading: false,
  loadingError: ''
});

const reducer = (state = initialState, action) => {
  const { payload } = action;

  switch(action.type) {
    case MODELS_PAGE_LOAD_REQUEST:
      return state.set('loading', true);
    case MODELS_PAGE_LOAD_ERROR:
      state.merge({ loadingError: payload, loading: false });
    case MODELS_PAGE_LOAD_SUCCESS:
      return state.set('loading', false);
    default:
      return state;
  }
}

export default reducer;
