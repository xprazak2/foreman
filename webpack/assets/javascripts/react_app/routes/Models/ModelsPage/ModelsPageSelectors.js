export const modelsPageSearchString = state => state.autocomplete.searchBar.searchQuery;
