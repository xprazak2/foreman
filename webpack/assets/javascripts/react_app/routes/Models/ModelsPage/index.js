import { connect } from 'react-redux';
import { compose, bindActionCreators } from 'redux';
import Immutable from 'seamless-immutable';

import ModelsPage from './ModelsPage';
import * as actions from './ModelsPageActions';

import { callOnMount, callOnPopState } from '../../../common/HOC';

import withDataReducer from '../../common/reducerHOC/withDataReducer';

import {
  selectModels,
  selectPage,
  selectPerPage,
  selectSearch,
  selectSort,
  selectHasData,
  selectHasError,
  selectIsLoading,
  selectSubtotal,
  selectMessage,
} from './ModelsPageSelectors';

const mapStateToProps = state => ({
  models: selectModels(state),
  page: selectPage(state),
  perPage: selectPerPage(state),
  search: selectSearch(state),
  sort: selectSort(state),
  isLoading: selectIsLoading(state),
  hasData: selectHasData(state),
  hasError: selectHasError(state),
  itemCount: selectSubtotal(state),
  message: selectMessage(state),
});

export const initialState = Immutable({
  total: 0,
  subtotal: 0,
  page: null,
  perPage: null,
  search: '',
  canCreate: false,
  sort: {
    by: null,
    order: null,
  },
  results: [],
});

export const reducers = {
  modelsPage: withDataReducer('MODELS_PAGE', initialState),
};

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  callOnMount(({ initializeModels }) => initializeModels()),
  callOnPopState(({ initializeModels }) => initializeModels())
)(ModelsPage);
