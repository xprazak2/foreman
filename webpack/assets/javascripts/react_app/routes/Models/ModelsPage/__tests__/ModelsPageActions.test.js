import { testActionSnapshotWithFixtures } from '@theforeman/test';

import { API } from '../../../../redux/API';

import {
  fetchModels,
  fetchAndPush,
  initializeModels,
} from '../ModelsPageActions';

import { propsFactory, models } from './ModelsPage.fixtures';

const fetchActionParams = {
  page: 1,
  perPage: 20,
  searchQuery: '',
  sort: { by: '', order: '' },
};

const response = propsFactory({
  results: models,
  total: 2,
  subtotal: 2,
  page: 1,
  perPage: 20,
});

const responseMock = { data: response };

jest.mock('../../../../redux/API/API');

const runWithGetState = (state, action, ...params) => dispatch => {
  const getState = () => ({
    modelsPage: state,
  });

  action(...params)(dispatch, getState);
};

const runApiFetch = (state, actionParams, serverMock) => {
  API.get.mockImplementation(serverMock);

  return runWithGetState(state, fetchModels, actionParams);
};

const fixtures = {
  'should fetch models': () =>
    runApiFetch(propsFactory(), fetchActionParams, async () => responseMock),

  'should fetch and fail': () =>
    runApiFetch(propsFactory(), fetchActionParams, async () => {
      const error = new Error();
      error.response = { status: 'KO', statusText: 'random-fail' };
      throw error;
    }),

  'should initialize models': () =>
    runWithGetState(propsFactory(), initializeModels, responseMock),

  'should fetch and push': () =>
    runWithGetState(propsFactory(), fetchAndPush, response),
};

describe('ModelsPage actions', () => testActionSnapshotWithFixtures(fixtures));
