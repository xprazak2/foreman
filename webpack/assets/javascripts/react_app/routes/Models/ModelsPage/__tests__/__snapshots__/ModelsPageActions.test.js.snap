// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelsPage actions should fetch and fail 1`] = `
Array [
  Array [
    Object {
      "type": "MODELS_PAGE_SHOW_LOADING",
    },
  ],
]
`;

exports[`ModelsPage actions should fetch and push 1`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`ModelsPage actions should fetch models 1`] = `
Array [
  Array [
    Object {
      "type": "MODELS_PAGE_SHOW_LOADING",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "canCreate": false,
        "hasData": true,
        "hasError": false,
        "isLoading": true,
        "message": Object {
          "text": "",
          "type": "empty",
        },
        "page": 1,
        "perPage": 20,
        "results": Array [
          Object {
            "canDelete": true,
            "canEdit": true,
            "hostsCount": 5,
            "id": 1,
            "name": "my-hw-model",
            "vendorClass": "custom",
          },
          Object {
            "canDelete": false,
            "canEdit": false,
            "hostsCount": 4,
            "id": 2,
            "name": "your-hw-model",
            "vendorClass": "B+",
          },
        ],
        "search": "",
        "sort": Object {
          "by": "",
          "order": "",
        },
        "subtotal": 2,
        "total": 2,
      },
      "type": "MODELS_PAGE_DATA_RESOLVED",
    },
  ],
  Array [
    Object {
      "type": "MODELS_PAGE_HIDE_LOADING",
    },
  ],
]
`;

exports[`ModelsPage actions should initialize models 1`] = `
Array [
  Array [
    [Function],
  ],
]
`;
